keps/sig-node/20191211-node-maintenance-lease.md

```diff
@@ -0,0 +1,175 @@
+  - [Use existing system namespace for lease object](#use-existing-system-namespace-for-lease-object)
+  - [Use optional non-system namespace](#use-optional-non-system-namespace)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+Node Maintenance Leases are useful to coordinate disruptive actions
+controllers or administrators might wish to take against a node. This lease
+is a voluntary mechanism and allows coordination among different components
+without those components needing knowledge of each other.
+
+## Motivation
+
+Currently, there is no centralized way to inform a controller or user that
+another controller or user is performing a disruptive operation against a node.
+Examples of disruptive operations include:
+1. Rebooting a node
+1. Powering off a node for maintenance
+1. Draining or Cordoning a node
+1. Deleting/deprovisioning a node due to failed health checks
+
+### Goals
+
+1. Define a central point of coordination for various actors.
+1. Define high-level steps for well-behaved clients to implement.
+
+### Non-Goals
+
+Define actors which might utilize the lease.
+
+## Proposal
+
+Utilize the existing `Lease` built-in API in the API group `coordination.k8s.io`.
+Create a new Lease object per-node  with Name equal to Node name in a newly
+created dedicated namespace
+That namespace should be created automatically (similarly to "default" and
+"kube-system", probably by NodeController) and never be deleted (so that nodes
+don't require permission for it).
+
+The Lease object should be created automatically, and the ownerRef should
+be the corresponding node so it is removed when the node is deleted.
+
+### User Stories [optional]
+
+#### Story 1
+
+I have a configuration management system that requires the node to be online to perform configuration updates.
+
+I have an automated-snapshot system that stops an instance and takes a snapshot for backup / data retention purposes.
+
+I do not want snapshots to happen during an update.
+
+#### Story 2
+
+I am investigating a problem with the node. I don't want the node to be disrupted
+by automation while I'm performing my investigation or fixes, and I don't want
+to stop the automation components from acting on other nodes.
+
```

Comment:
Another story: an automated reboot tool https://github.com/weaveworks/kured/ wants to flag that the machine is being rebooted; also wants to avoid rebooting a machine that some other tool or person is currently doing maintenance on.

Reply:
@bboreham cluster-api is also working on something similar around MachineHealthChecks.  Definite a broad usecase.